strawberryfields/io/utils.py

```diff
@@ -0,0 +1,162 @@
+        code_seq.append(f"prog = sf.Program({prog.num_subsystems})")
+
+    # check if an engine is supplied; if so, format and add backend/target
+    # along with backend options
+    if eng:
+        if hasattr(eng, "connection"):
+            code_seq.append(f'eng = sf.RemoteEngine("{eng.target}")')
+        else:
+            if "cutoff_dim" in eng.backend_options:
+                formatting_str = (
+                    f'"{eng.backend_name}", backend_options='
+                    + f'{{"cutoff_dim": {eng.backend_options["cutoff_dim"]}}}'
+                )
+                code_seq.append(f"eng = sf.Engine({formatting_str})")
+            else:
+                code_seq.append(f'eng = sf.Engine("{eng.backend_name}")')
+
+    # check if program is a TDMProgram and format the context as appropriate
+    if isinstance(prog, TDMProgram):
+        # if the context arrays contain pi-values, factor out multiples of np.pi
+        tdm_params = [f"[{_factor_out_pi(par)}]" for par in prog.tdm_params]
+        code_seq.append("\nwith prog.context(" + ", ".join(tdm_params) + ") as (p, q):")
+    else:
+        code_seq.append("\nwith prog.context as q:")
+
+    # add the operations, and replace any free parameters with e.g. `p[0]`, `p[1]`
+    for cmd in prog.circuit or []:
+        name = cmd.op.__class__.__name__
+        if isinstance(prog, TDMProgram):
+            format_dict = {k: f"p[{k[1:]}]" for k in prog.parameters.keys()}
+            params_str = _factor_out_pi(cmd.op.p).format(**format_dict)
+        else:
+            params_str = _factor_out_pi(cmd.op.p)
+
+        modes = [f"q[{r.ind}]" for r in cmd.reg]
+        if len(modes) == 1:
+            modes_str = ", ".join(modes)
+        else:
+            modes_str = "(" + ", ".join(modes) + ")"
+        op = f"    ops.{name}({params_str}) | {modes_str}"
+
+        code_seq.append(op)
+
+    if eng:
+        code_seq.append("\nresults = eng.run(prog)")
+
+    return "\n".join(code_seq)
+
+
+def _factor_out_pi(num_list: List[Union[Number, str]], denominator: int = 12) -> str:
+    """Factors out pi, divided by the denominator value, from all number in a list
+    and returns a string representation.
+
+    Args:
+        num_list (list[Number, str]): a list of numbers and/or strings
+        denominator (int): factor out pi divided by denominator;
+            e.g. default would be to factor out np.pi/12
+
+    Return:
+        string: containing strings of values and/or input string objects
+    """
+    factor = np.pi / denominator
+
+    a = []
+    for p in num_list:
+        # if list element is not a number then append its string representation
+        if not isinstance(p, (int, float)):
+            a.append(str(p))
+            continue
+
+        if np.isclose(p % factor, [0, factor]).any() and p != 0:
+            gcd = np.gcd(int(p / factor), denominator)
+            if gcd == denominator:
+                if int(p / np.pi) == 1:
+                    a.append("np.pi")
```

Comment:
Will the XIR expect these to just be `pi`? Or is that irrelevant here?
```suggestion
                    a.append("pi")
```

Reply:
So, this has nothing to do with XIR. It's only used in the `generate_code` function (thus private), which generates SF code, which uses NumPy. So it should be OK.

Reply:
Ah, sounds good!